### PR TITLE
Noise simulations overhaul

### PR DESCRIPTION
**Pull request by msyriac**
Ready for review. The objectives are:

- [x] 1. intra-tube covariance support for both CAR and healpix
- [x] 2. reprojection of v0.2 templates to CAR
- [x] 3. solve the tube noise scaling issue #43 
- [x] 4. support homogenous sims (i.e. no scan strategy, full-sky sim) #40 
- [x] 5. support white noise (i.e. turning off atmosphere)
- [ ] ~~6. support re-scaling of the white noise level (e.g. to use for ACT-like sims, or white noise Planck-like sims when combined with the previous feature)~~
- [ ] ~~7. produce maps of the inverse variance in each pixel from the hitmap and white noise level (a standard ACT data product) https://github.com/simonsobs/map_based_simulations/issues/7~~
- [x] 8. cache hitmaps -- no need to re-load or reproject for each realization
- [x] 9. get beam FWHMS, since many use cases require non-beam-deconvolved sims, we should provide an interface that tells us what the noise curve code uses for the beam FWHM
- [x] 10. use a tuple-based seed system to not worry about seed collisions
- [x] 11. solve some bugs in the CAR implementation #38 
- [x] 12. Set rolloff_ell default to 50 #42

### CODE REVIEW

**Comment by msyriac** [comment]
Ready for review now. The white noise scaling and inverse variance maps will be in a separate PR. The tests for Travis need to be regenerated since this is not backwards compatible. I'm also not familiar with the modifications needed for runner.py. I will continue to run some tests of the power spectra of the generated sims in the meantime.

**Comment by msyriac** [comment]
Unit tests are in and passing locally. pixell (on PyPI) requirement has been added.

**Comment by zonca** [comment]
will review it later this week

File: mapsims/noise.py [hunk 1]
```diff
@@ -116,103 +149,152 @@ def __init__(
...
                 ell_max if ell_max is not None else 10000 * (1.0 / self._pixheight)
             )
+            self.pmap = enmap.pixsizemap(self.shape, self.wcs)
         else:
             assert shape is None
             assert wcs is None
             self.healpix = True
             self.nside = nside
             self.ell_max = ell_max if ell_max is not None else 3 * nside
+            self.pmap = 4*np.pi / hp.nside2npix(nside)
```
[the remark this answers is not in the document]
Comment (msyriac) [reply mapsims/noise.py]: Done.

File: mapsims/noise.py [hunk 2]
```diff
@@ -116,103 +149,152 @@ def __init__(
...
         SA_one_over_f_mode : {"pessimistic", "optimistic", "none"}
             Correlated noise performance of the detectors on the Small Aperture telescopes
-        hitmap_version : string
-            Version string for hitmaps stored remotely.
+        sky_fraction : optional,float
+            If homogenous is True, this sky_fraction is used for the noise curves.
+        cache_hitmaps : bool
+            If True, caches hitmaps.
+        boolean_sky_fraction: bool
+            If True, determines fsky based on fraction of hitmap that is zero. If False,
```
[the remark this answers is not in the document]
Comment (msyriac) [reply mapsims/noise.py]: Done in API. Internally, some variables are called fsky.

File: mapsims/noise.py [hunk 3]
```diff
@@ -116,103 +149,152 @@ def __init__(
...
+        hitmaps. Each loaded `telescope` is kept in memory, but
+        new choice of `scanning_strategy` erases the previous one.
+
+        Parameters
+        ----------
+
+        telescope : string
+            Telescope identifier, typically `LA` or `SA` for the large aperture
+            and small aperture, respectively.
+
```
[the remark this answers is not in the document]
Comment (msyriac) [reply mapsims/noise.py:287]: Done.

File: mapsims/noise.py [hunk 4]
```diff
@@ -116,103 +149,152 @@ def __init__(
...
+        telescope = f"{tube[0]}A"  # get LA or SA from tube name
+        survey = self._get_survey(tube)
+        if telescope == "SA":
             ell, noise_ell_T, noise_ell_P = survey.get_noise_curves(
-                1.0,  # We load hitmaps later, so we compute and apply sky fraction later
+                ncurve_fsky,  # We load hitmaps later, so we compute and apply sky fraction later
                 self.ell_max,
                 delta_ell=1,
-                full_covar=self.full_covariance,
+                full_covar=True,
```
[the remark this answers is not in the document]
Comment (msyriac) [reply mapsims/noise.py]: I wasn't sure if the shape of the arrays returned by the noise code changes when that argument is changed. To protect against that and make the code simpler, I always ask for full covariance, but set the correlation to zero later on depending on the value of full_covariance.

File: mapsims/noise.py [hunk 5]
```diff
@@ -8,45 +8,89 @@
...
+    """
+    tubes = so_utils.tubes
+    bands = tubes[tube]
+    freqs = [so_utils.band_freqs[x] for x in bands]
+    available_frequencies = np.unique(so_utils.frequencies)
+    band_ids = [available_frequencies.searchsorted(f) for f in freqs]
+    return band_ids
+
+
+def _band_index(tube, band):
```
[the remark this answers is not in the document]
Comment (msyriac) [reply mapsims/noise.py:52]: Agree, that might be better. I'll look into this.

File: mapsims/noise.py [hunk 6]
```diff
@@ -116,103 +149,152 @@ def __init__(
...
             assert wcs is not None
             self.healpix = False
-            self.shape = shape
+            self.shape = shape[-2:]
             self.wcs = wcs
             self._pixheight = np.abs(wcs.wcs.cdelt[0] * 60.0)
             self.ell_max = (
                 ell_max if ell_max is not None else 10000 * (1.0 / self._pixheight)
             )
+            self.pmap = enmap.pixsizemap(self.shape, self.wcs)
```
[the remark this answers is not in the document]
Comment (msyriac) [reply mapsims/noise.py]: Done.

File: mapsims/noise.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (msyriac) [reply mapsims/noise.py:52]: Hmm, I looked at this a bit more and now am not sure there's actually anything we can improve. _band_ids_from_tube is something that is needed in order to convert a tube specification to the SO noise code's ordering. If we were going to load stuff into a dictionary, we would have to use this sort of function anyway. All that _band_ids_from_tube does is encapsulate the scheme you were using in your original code to index stuff, but place it in a function that can be reused (e.g. for getting beam FWHM or white noise) to reduce code duplication.  _band_index serves a similar purpose -- it's just helping with the ordering that already exists in so_utils. If we were to add an intermediate dictionary, we would still have to use a function like _band_index to populate. Hopefully the doctests I've added will prevent any future errors.

[the remark this answers is not in the document]
Comment (zonca) [reply mapsims/noise.py:52]: ok

**Comment by zonca** [comment]
very good, the noise unit tests pass also on my machine, now I'll update the rest of the package to your changes and then fix Travis

**Comment by zonca** [comment]
@msyriac can you please double-check https://github.com/simonsobs/mapsims/pull/44/commits/2ba401d615c75f37905c8a048079b5c444c2c57e?

**Comment by msyriac** [comment]
> @msyriac can you please double-check [2ba401d](https://github.com/simonsobs/mapsims/commit/2ba401d615c75f37905c8a048079b5c444c2c57e)?

Looks good, thank you.

**Comment by zonca** [comment]
ok, merging this, another open issue is to expose the car configuration to `MapSim`, but that is easier to work on when I am actually preparing a release.

**Pull request merged**